src/aioswitcher/api/__init__.py

```diff
@@ -162,20 +171,33 @@ async def _login(self) -> Tuple[str, SwitcherLoginResponse]:
 
         """
         timestamp = current_timestamp_to_hexadecimal()
-        if (
+        if self._device_type and self._device_type.token_needed and bool(self._token):
+            packet = packets.LOGIN_TOKEN_PACKET_TYPE2.format(
+                self._token, timestamp, self._device_id
+            )
+        elif (
             self._device_type
             and self._device_type == DeviceType.BREEZE
             or self._device_type == DeviceType.RUNNER
             or self._device_type == DeviceType.RUNNER_MINI
         ):
-            packet = packets.LOGIN2_PACKET_TYPE2.format(timestamp, self._device_id)
+            packet = packets.LOGIN_PACKET_TYPE2.format(timestamp, self._device_id)
         else:
             packet = packets.LOGIN_PACKET_TYPE1.format(timestamp, self._device_key)
         signed_packet = sign_packet_with_crc_key(packet)
 
         logger.debug("sending a login packet")
         self._writer.write(unhexlify(signed_packet))
         response = await self._reader.read(1024)
+
+        if self._device_type and self._device_type.token_needed and bool(self._token):
```

Comment:
There is only the 1 now in line 179.. this code is outdated..

Reply:
Thanks, that's better. But I wonder if an Enum can be empty.

Reply:
An example would be better. Consider the following code:

```python
from enum import Enum


class MyEnum(Enum):
    Member1 = 1
    Member2 = 2


class MyClass:
    def __init__(self, member: MyEnum):
        self._member = member

    def member(self) -> MyEnum:
        return self._member


if __name__ == "main":
    print(MyClass(None).member())
```

When you lint this with _MyPy_, you get the following error:

```
error: Argument 1 to "MyClass" has incompatible type "None"; expected "MyEnum"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

So our _MyPy_ CI will fail if a contribution includes passing `None` as the device type.

In our documentation, we instruct users to use _MyEnum_ when instantiating _MyClass_. If they don't adhere, the code should either break or act unexpectedly.
---

Your contribution includes the following:

```python
elif (
            self._device_type
            and self._device_type == DeviceType.BREEZE
            or self._device_type == DeviceType.RUNNER
            or self._device_type == DeviceType.RUNNER_MINI
        ):
```

My question was: Why are you evaluating `self._device_type`? If the device type is `None,` that means the user deliberately instantiated the API object that way.
This also means the next statement `self._device_type == DeviceType.BREEZE` will break the code, which is exactly what I think we should do! Break the code if the device type is `None`.

- Do we have any supported use cases in our code where we don't need to know the device type?
- Do we have a "default" device type?

If the answer to these questions is no, then not only should this statement be removed, but we should also consider verifying that the device type at the init function and throwing an error if needed.

@thecode What do you think?

Reply:
When we started we had only one type of API (so all devices used the same type), later other types added (so most devices fall into default types) and now most calls are depended on the type of the device, reviewing the state of the implementation now it can easily break if type is not supplied I would just suggest to make it a mandatory.

I suggest to expose only one API init function (in future PR, not this one maybe) and call the different APIs internally by the type of the device, so the user doesn't have to create `SwitcherType1Api`/`SwitcherType2Api` according to the device type, but it will only create a single class - `SwitcherApi` which will internally call the correct logic/messages according to the device type.

Reply:
It is done.